tests/devices/qubit/test_simulate.py

```diff
@@ -514,6 +514,39 @@ def test_multi_measurements(self):
 
         assert result[2].shape == (10000, 2)
 
+    def test_shots_reuse(self, mocker):
+        """Test that samples are reused when two measurements commute"""
+        ops = [qml.Hadamard(0), qml.CNOT([0, 1])]
+        mps = [
+            qml.expval(qml.PauliX(0)),
+            qml.expval(qml.PauliX(1)),
+            qml.expval(qml.PauliZ(0)),
+            qml.var(qml.PauliX(1)),
+            qml.var(qml.PauliY(0)),
+            qml.probs(wires=[0]),
+            qml.probs(wires=[0, 1]),
+            qml.sample(wires=[0, 1]),
+            qml.expval(
+                qml.Hamiltonian([1.0, 2.0, 3.0], [qml.PauliX(0), qml.PauliZ(1), qml.PauliY(1)])
+            ),
+            qml.expval(qml.sum(qml.PauliX(0), qml.PauliZ(1), qml.PauliY(1))),
+            qml.expval(qml.s_prod(2.0, qml.PauliX(0))),
+            qml.expval(qml.prod(qml.PauliX(0), qml.PauliY(1))),
+        ]
+
+        qs = qml.tape.QuantumScript(ops, mps, shots=100)
+
+        spy = mocker.spy(qml.devices.qubit.sampling, "sample_state")
```

Comment:
I'll try and find a way for us to "track" this kind of information. Then we can check the `tracker` and not have to mock, but for now mocking should be fine.